oeps/oep-0012.rst

```diff
@@ -0,0 +1,475 @@
+  choose to show a different UI for its "Course" tab. As another example,
+  a mobile runtime could show native mobile versions of certain blocks,
+  perhaps using `React Native`_.
+
+* XBlocks could use pre-existing component views rather than rolling their
+  own. A good example would be that a standard "Settings" component view could
+  be provided that any XBlock could use as its authoring view. Another example
+  would be rewriting the Inline Discussion XBlock to just render the same
+  component view as is being used by the "Discussion" tab.
+
+* Developers could subclass a component view to customize its behavior without
+  having to subclass the XBlock itself.
+
+There are some reasons why it may be better to keep XBlocks as they are:
+
+* It requires XBlock authors to use and understand Django.
+
+* It is more complex if even simple XBlocks must declare a separate Django
+  component view.
+
+Note that there is nothing stopping an XBlock author from including both
+XBlock and Django in their project, and then calling out to the component
+view from their XBlock's `student_view` method. For example, this is almost
+certainly what we will do for discussions so that the same component can be
+rendered for the XBlock, the discussion tab, and for team discussions.
+
+Provide isolation of Django Component Views
+-------------------------------------------
+
+There are known challenges with supporting third-party XBlocks which will
+also apply to Django Component Views:
+
+* When a page is synchronously rendered, a poorly performing block/component
+  will prevent the rest of the page from appearing.
+
+* The CSS provided by the block/component is not scoped and so can affect
+  the styling of other sections of the page.
+
+* The JavaScript runs in a global context and so malicious and/or poorly
+  written code can affect the behavior of the whole page.
+
+Several approaches have been attempted to reduce these risks for XBlocks,
+and there are promising possibilities with the use of web components. This
+is an area that edX will continue to explore, but in the meantime it should
+not be a blocker for providing pluggable user interfaces.
+
+Note that it is already a policy to only install trusted Django apps, because
+there are so many ways in which they could be dangerous beyond those listed
+here. The addition of component views to an otherwise trusted app should not add
+much in the way of additional review required to be confident in the behavior
+of the app.
+
+Consider how web components and Django Component Views overlap
+--------------------------------------------------------------
+
+There is a lot of interest in the `Web Components`_ specification and how
+the capabilities could be used to provide reusable components. In particular,
+it seems that web components provide a level of isolation that XBlocks do not,
+and that which could greatly enhance the security of the edX platform.
```

Comment:
@bjacobel great point. I've removed this sentence since it is inaccurate, and I've replaced it with the following:

> More exploration is
needed to determine whether web components are ready for use within Open edX,
and if so how they overlap with Django Component Views.